### PR TITLE
Incremental Janusgraph schema updates #1588

### LINKED ISSUE #1588: Incremental Janusgraph schema updates
# Task Description
We need to be able to manage Janusgraph schema with incremental versioned updates. This should be done with help of `JanusGraphManagement` which provide Java API to manage schema. We already have a `org.carlspring.strongbox.janusgraph.graph.schema.StrongboxSchema` class that creates initial schema and within this task we should make it work using step-by-step change-sets. The change-sets will be stored as a constant list of `StrongboxSchema` class.

# Tasks

The following tasks will need to be carried out:
* [ ] Implement an interface which will represent single change-set, it should have the `apply(JanusGraphManagement jgm)` method and following attributes:
- name (some meanable change-set name)
- version (Semantic Versioning format)
* [ ] Add `DatabaseSchema` vertex type which will represent current schema
* [ ] Add `ChangeSet` vertex type which will represent applied change-sets
* [ ] Add `DatabaseSchema#ChangeSet` edge type which will represent one-to-many relation between `DatabaseSchema` and `ChangeSet` (it should also have the `applyDate` property)
* [ ] Implement the logic which will create `DatabaseSchema` vertex if it not exists
* [ ] Implement the logic which will apply change-sets delta from current `DatabaseSchema` version.
* [ ] implement `StrongboxSchema.dropSchema` method for test purpose which will be executed on Spring Application context shutdown.

# Task Relationships

This task:
* Is a sub-task of: #1573

# Useful Links

* [Schema and Data Modeling](https://docs.janusgraph.org/basics/schema/)

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @fuss86
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by sbespalov**
Closes [#1588](https://github.com/strongbox/strongbox/issues/1588)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+[strongbox](https://github.com/strongbox/strongbox) uses this project submodules as required dependencies. During the startup of [Strongbox](https://github.com/strongbox/strongbox), the application detects if an JanusGraph database already exists. If not, then the application uses an extracted empty JanusGraph database snapshot from the `strongbox-db-import` artifact. If the JanusGraph database already exists, then the application applies all missing changesets from the `strongbox-db-schema` artifact.
 
 ## How to build this project ?
 To build the code, simply execute:
 `mvn clean install`
 
 ## What's the result of the build process ?
 This project produces the following artifacts:
-* `strongbox-db-liquibase-${version}.jar` : Located in `strongbox-db-liquibase/target`, which contains current database liquibase changesets
+* `strongbox-db-shcema-${version}.jar` : Located in `strongbox-db-schema/target`, which contains current database changesets
```
Comment (carlspring) [inline README.md]:
```suggestion
* `strongbox-db-schema-${version}.jar` : Located in `strongbox-db-schema/target`, which contains current database changesets
```

File: README.md [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
 
 ## How is this project used in [strongbox](https://github.com/strongbox/strongbox) ?
-[strongbox](https://github.com/strongbox/strongbox) uses this project submodules as required dependencies. During the startup of [Strongbox](https://github.com/strongbox/strongbox), the application detects if an OrientDB database already exists. If not, then the application uses an extracted empty OrientDB database snapshot from the `strongbox-db-import` artifact. If the OrientDB database already exists, then the application applies all missing liquibase changesets from the `strongbox-db-liquibase` artifact.
+[strongbox](https://github.com/strongbox/strongbox) uses this project submodules as required dependencies. During the startup of [Strongbox](https://github.com/strongbox/strongbox), the application detects if an JanusGraph database already exists. If not, then the application uses an extracted empty JanusGraph database snapshot from the `strongbox-db-import` artifact. If the JanusGraph database already exists, then the application applies all missing changesets from the `strongbox-db-schema` artifact.
```
Comment (carlspring) [inline README.md]:
```suggestion
[strongbox](https://github.com/strongbox/strongbox) uses this project submodules as required dependencies. During the startup of [Strongbox](https://github.com/strongbox/strongbox), the application detects if a JanusGraph database already exists. If one doesn't exist, it uses an extracted empty JanusGraph database snapshot from the `strongbox-db-import` artifact; if a database exists, then the application applies all missing changesets from the `strongbox-db-schema` artifact.
```

File: strongbox-db-import/src/main/java/org/carlspring/strongbox/dbimport/Application.java [hunk 3]
```diff
@@ -1,19 +1,23 @@
-package org.carlspring.strongbox.db;
+package org.carlspring.strongbox.dbimport;
```
Comment (carlspring) [inline strongbox-db-import/src/main/java/org/carlspring/strongbox/dbimport/Application.java]: Cosmetic: Perhaps a better package name would be `org.carlspring.strongbox.db.import`?

File: strongbox-db-schema/src/main/java/org/carlspring/strongbox/db/schema/Properties.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package org.carlspring.strongbox.db.schema;
+
+/**
+ * @author ankit.tomar
+ */
+public interface Properties
```
Comment (carlspring) [inline strongbox-db-schema/src/main/java/org/carlspring/strongbox/db/schema/Properties.java]: Why are all these constants in an interface? A lot of them are layout provider-specific. Wouldn't it be better to have at least part of them in the respective modules, under a package like `org.carlspring.strongbox.db.properties` and then have a class like, for example, `Maven2LayoutProviderProperties`?

What do you think, @sbespalov , @anki2189 , @steve-todorov ?

File: strongbox-db-import/src/main/java/org/carlspring/strongbox/dbimport/Application.java (hunk 3 above)
Comment (sbespalov) [reply strongbox-db-import/src/main/java/org/carlspring/strongbox/dbimport/Application.java]: this `dbimport` package name prevents Spring to autoscan `EmbeddedDbServerConfiguration` from `strongbox-data-service`.

File: strongbox-db-schema/src/main/java/org/carlspring/strongbox/db/schema/Properties.java (hunk 4 above)
Comment (sbespalov) [reply strongbox-db-schema/src/main/java/org/carlspring/strongbox/db/schema/Properties.java]: @carlspring  by "respective modules" you mean `strongbox-storage-layout-providers` modules?

File: strongbox-db-import/src/main/java/org/carlspring/strongbox/dbimport/Application.java (hunk 3 above)
Comment (carlspring) [reply strongbox-db-import/src/main/java/org/carlspring/strongbox/dbimport/Application.java]: Thanks for clarifying! :)

File: strongbox-db-schema/src/main/java/org/carlspring/strongbox/db/schema/Properties.java (hunk 4 above)
Comment (carlspring) [reply strongbox-db-schema/src/main/java/org/carlspring/strongbox/db/schema/Properties.java]: Yeah, the layout providers and perhaps (in the future) other modules as well.

Comment (sbespalov) [reply strongbox-db-schema/src/main/java/org/carlspring/strongbox/db/schema/Properties.java]: this is not possible, we couldn't access anything within `strongbox-storage-layout-providers` modules from `strongbox-db`.

File: README.md (hunk 1 above)
Comment (sbespalov) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (sbespalov) [reply README.md]: fixed

**Pull request merged**